lib/rouge/lexers/ttcn3.rb

```diff
@@ -0,0 +1,129 @@
+          oct2unichar str2int str2hex str2oct str2float enum2int
+          any2unistr lengthof sizeof ispresent ischosen isvalue isbound
+          istemplatekind regexp substr replace encvalue decvalue
+          encvalue_unichar decvalue_unichar encvalue_o decvalue_o
+          get_stringencoding remove_bom rnd hostid send receive
+          setverdict
+        )
+      end
+
+      def self.reserved
+        @reserved = %w(
+          all alt apply assert at configuration conjunct const control
+          delta deterministic disjunct duration fail finished fuzzy
+          history implies inconc inv lazy mod mode notinv now omit
+          onentry onexit par pass prev realtime seq setstate static
+          stepsize stream timestamp until values wait
+        )
+      end
+
+      def self.types
+        @types = %w(
+          anytype address boolean bitstring bytestring charstring
+          component enumerated float integer hexstring octetstring port
+          record set of union universal
+        )
+      end
+
+      # optional comment or whitespace
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+      const_name = /[A-Z][A-Z0-9_]*\b/
+      module_name = /[A-Z][a-zA-Z0-9]*\b/
+
+      state :root do
+        rule %r/[^\S\n]+/, Text
+
+        rule %r(//.*), Comment::Single
+        rule %r(/\*.*?\*/)m, Comment::Multiline
+
+        # keywords: go before method names to avoid lexing "throw new XYZ"
+        # as a method signature
+        rule %r(
+          (\s*(?:[a-zA-Z_][a-zA-Z0-9_.\[\]<>]*\s+)+?) # return arguments
+          ([a-zA-Z_][a-zA-Z0-9_]*)                  # method name
+          (\s*)(\()                                 # signature start
+        )mx do |m|
+          # TODO: do this better, this shouldn't need a delegation
+          delegate TTCN3, m[1]
+          token Name::Function, m[2]
+          token Text, m[3]
+          token Operator, m[4]
+        end
+
+        rule %r/(?:true|false|null)\b/, Keyword::Constant
+        rule %r/(module)\b/, Keyword::Declaration, :module
+        rule %r/import\b/, Keyword::Namespace, :import
+        rule const_name, Name::Constant
+        rule module_name, Name::Label
+
+        rule %r/(\.)(#{id})/ do
+          groups Operator, Name::Attribute
+        end
+        rule %r/@#{id}/, Name::Decorator
+        rule %r/#{id}:/, Name::Label
+        rule %r/\$#{id}/, Name
+
+        rule id do |m|
+          if self.class.keywords.include? m[0]
+            token Keyword
+          elsif self.class.reserved.include? m[0]
+            token Keyword::Reserved
+          elsif self.class.types.include? m[0]
+            token Keyword::Type
+          else
+            token Name
+          end
+        end
+
+        rule %r/"(\\\\|\\"|[^"])*"/, Str
+        rule %r/'(?:\\.|[^\\]|\\u\h{4})'/, Str::Char
+
+        rule %r/[~^*!%&<>\|+=\/?-]/, Operator
+        rule %r/[@#\$()`{}\[\]:;,.\\]/, Punctuation
```

Comment:
I rationalised these based on common sense but could you check whether the correct characters are in the correct rule? The original PR had a large number of duplicates.

Reply:
@YannGarcia Can you confirm the above rules look OK?

Reply:
Hello,
My apologies for the delay.
I did a full review of the TTCN-3 rules. I think it looks better now.